Code/GraphMol/GenericGroups/GenericGroups.cpp

```diff
@@ -0,0 +1,481 @@
+    if (aidx != static_cast<int>(atom.getIdx()) &&
+        (ignore[aidx] || (matcher && !matcher(*mol.getAtomWithIdx(aidx))))) {
+      return false;
+    }
+    if (!atLeast && atLeastOne(*mol.getAtomWithIdx(aidx))) {
+      atLeast = true;
+    }
+  }
+  return atLeast;
+}
+bool checkBondRing(const ROMol &mol, const std::vector<int> &bring,
+                   BondMatcherFunc matcher, BondMatcherFunc atLeastOne) {
+  bool atLeast = atLeastOne == nullptr;
+  for (auto bidx : bring) {
+    if (matcher && !matcher(*mol.getBondWithIdx(bidx))) {
+      return false;
+    }
+    if (!atLeast && atLeastOne(*mol.getBondWithIdx(bidx))) {
+      atLeast = true;
+    }
+  }
+  return atLeast;
+}
+
+bool FusedRingMatch(const ROMol &mol, const Atom &atom,
+                    boost::dynamic_bitset<> ignore,
+                    AtomMatcherFunc atomMatcher = nullptr,
+                    BondMatcherFunc bondMatcher = nullptr,
+                    AtomMatcherFunc atLeastOneAtomPerRing = nullptr,
+                    BondMatcherFunc atLeastOneBondPerRing = nullptr,
+                    AtomMatcherFunc atLeastOneAtom = nullptr) {
+  PRECONDITION(&atom.getOwningMol() == &mol, "atom not owned by molecule");
+  if (atomMatcher && !atomMatcher(atom)) {
+    return false;
+  }
+
+  if (!mol.getRingInfo() || !mol.getRingInfo()->isInitialized()) {
+    MolOps::findSSSR(mol);
+  }
+  if (!mol.getRingInfo()->numAtomRings(atom.getIdx())) {
+    return false;
+  }
+
+  // we're going to traverse over the entire fused ring system involving this
+  // atom start by finding the first ring:
+  std::set<int> ringAtoms;
+  for (auto i = 0u; i < mol.getRingInfo()->numRings(); ++i) {
+    const auto &ring = mol.getRingInfo()->atomRings()[i];
+    if (std::find(ring.begin(), ring.end(), atom.getIdx()) != ring.end()) {
+      if (!checkAtomRing(mol, atom, ignore, ring, atomMatcher,
+                         atLeastOneAtomPerRing)) {
+        return false;
+      }
+      if (!checkBondRing(mol, mol.getRingInfo()->bondRings()[i], bondMatcher,
+                         atLeastOneBondPerRing)) {
+        return false;
+      }
+
+      ringAtoms.insert(ring.begin(), ring.end());
+      break;
+    }
+  }
+  // now loop over all rings and find the ones which share at least two atoms
+  // with what we've seen so far
+  for (auto i = 0u; i < mol.getRingInfo()->numRings(); ++i) {
+    const auto &ring = mol.getRingInfo()->atomRings()[i];
+    // check overlap of this ring with what we've seen so far and make sure that
+    // the new atoms we are adding all pass the test
+    std::set<int> sring(ring.begin(), ring.end());
+    std::vector<int> diff(sring.size());
+    auto dit =
+        std::set_difference(sring.begin(), sring.end(), ringAtoms.begin(),
+                            ringAtoms.end(), diff.begin());
+    auto numNewAtoms = dit - diff.begin();
+    if (!numNewAtoms || sring.size() - numNewAtoms < 2) {
+      // we don't overlap by at least two atoms
+      continue;
+    }
```

Comment:
I don't really understand this too well.

 Couldn't we instead use `std::set_intersection()`, to get exactly the atoms that overlap, and then just check the size of the result?

Reply:
That would be simpler at this point in the code, but at line 292 we need to update `ringAtoms` with the new atoms. 
It could be an unnecessary optimization, but it seems more efficient to do that with only the atoms which are not shared (what the current code does):
```
    ringAtoms.insert(diff.begin(), dit);
```
instead of using the entire ring, which I think would be necessary with your suggestion:
```
    ringAtoms.insert(sring.begin(), sring.end());
```

I'm happy to go either way

Reply:
Ok, it makes sense to me as it is.

`std::set_intersection()` and `ringAtoms.insert(sring.begin(), sring.end());` may be a bit more readable, but I hadn't considered performance, so I was wondering if there was something I was missing.